### ISSUE TITLE
Don't terminate when last client leaves

### ISSUE THREAD

**Issue by mweinelt** [owenthereal/upterm#82]
When the last client leaves the upterm session my upterm host session terminates. This is not ideal, as I could be running an editor and have unsaved changes.

As soon as the last user terminates their session by closing their terminal this is shown:

```
[...]
 ~ ▓▒░ Error: EOF
Usage:
  upterm host [flags]

Examples:
  # Host a terminal session that runs $SHELL with
  # client's input/output attaching to the host's
  upterm host

  # Host a terminal session that only allows specified public key(s) to connect
  $ upterm host --authorized-key PATH_TO_PUBLIC_KEY

  # Host a session with a custom command.
  upterm host -- docker run --rm -ti ubuntu bash

  # Host a session that runs 'tmux new -t pair-programming' and
  # force clients to join with 'tmux attach -t pair-programming'.
  # This is similar to tmate.
  upterm host --force-command 'tmux attach -t pair-programming' -- tmux new -t pair-programming

  # Use a different Uptermd server and host a session via WebSocket
  upterm host --server wss://YOUR_UPTERMD_SERVER -- YOUR_COMMAND

Flags:
  -a, --authorized-key string   an authorized_keys file that lists public keys that are permitted to connect.
  -f, --force-command string    force execution of a command and attach its input/output to client's.
  -h, --help                    help for host
      --known-hosts string      a file contains the known keys for remote hosts (required). (default "/home/hexa/.ssh/known_hosts")
  -i, --private-key strings     private key for public key authentication against the upterm server (required). (default [/home/hexa/.ssh/id_ed25519,/home/hexa/.ssh/id_rsa])
  -r, --read-only               host a read-only session. Clients won't be able to interact.
      --server string           upterm server address (required), supported protocols are shh, ws, or wss. (default "ssh://uptermd.upterm.dev:22")

FATA[0031] EOF   
```

**Comment by owenthereal** [comment]
@mweinelt:

What command did you run? For example, if you run `bash` (e.g. `upterm host -- bash`) and the client types `exit`, the `exit` is sent to the `bash` command and it will exit the bash process on the host. If you want a quitting client that doesn't exit your host running process, you would need a terminal multiplexer like `tmux`. `tmux` runs a server on the host and quitting client won't exit the tmux server on the host. For example to create a tmux session on the host and force clients to join by attaching to the tmux server: `upterm host --force-command 'tmux attach -t pair-programming' -- tmux new -t pair-programming` (Ref: https://github.com/owenthereal/upterm#quick-reference)

**Comment by mweinelt** [comment]
Right, i was running just plain zsh. Somewhat unexpected but understandable.

**Comment by mweinelt** [comment]
Okay, to clarify, I was in a read-only session. So I'd still call this unexpected.

```
upterm host --read-only zsh
```

**Comment by m0sth8** [comment]
I think I get what is the issue here. It works only when `--force-command` isn't used. It also works for --read-only sessions.

How to reproduce:
1. Run a host terminal `upterm host --read-only` in a terminal window.
2. Run a client terminal `ssh ..@uptermd.upterm.dev` in another terminal window.
3. Terminate (close) the client terminal. 
Expected: The host terminal do nothing dramatic (triggers ClientLeftCallback callback)
Actual: The host terminal exited with EOF error.

The reason for this behavior is the next few issues:
1. Write and Read-only client terminals affects host terminal dimensions. issue #138
2. When client terminal is terminated, it sends an event with width=0 and height=0 to host terminal. (https://github.com/owenthereal/upterm/blob/master/host/internal/server.go#L253). 
3. That cause host terminal to redraw.
5. Redraw triggers a write operation to writers https://github.com/owenthereal/upterm/blob/master/host/internal/command.go#L127. It has 3 writers at this point (MultiWriter internal cache, host terminal stdout, remote session). 
6. Because remote session is already terminated that causes EOF error. 

Possible ways to fix:
1. Don't redraw the host terminal when the client terminal gets killed.
2. Check session.Context().Done() somewhere (e.g.  https://github.com/owenthereal/upterm/blob/master/host/internal/server.go#L195 ) that would trigger a dead writer to be removed here https://github.com/owenthereal/upterm/blob/master/host/internal/server.go#L242
3. Handle writes to clients differently and ignore if they're broken/unstable. I believe that might also fix https://github.com/owenthereal/upterm/issues/134 but I didn't investigate that case.

**Comment by horazont** [comment]
This seems to not only happen when the last client leaves.

We can reproducibly bring upterm to terminate when *any* client of a read-only session closes (either with the SSH sequence `[enter] ~ .`, or by closing the terminal window).
